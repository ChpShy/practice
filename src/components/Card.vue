<script setup lang="ts">
import type { PropType } from 'vue'

interface ContactInfo {
  icon: string
  value: string
  desc: string
}

interface SocialInfo {
  icon: string
  link: string
}

defineProps({
  width: {
    type: String,
    default: '15rem',
  },
  avatar: {
    type: String,
  },
  userName: {
    type: String,
  },
  userRole: {
    type: String,
  },
  wrapClass: {
    type: [String, Array, Object],
  },
  headerClass: {
    type: [String, Array, Object],
  },
  contactClass: {
    type: [String, Array, Object],
  },
  footerClass: {
    type: [String, Array, Object],
  },
  contactList: {
    type: Array as PropType<ContactInfo[]>,
  },
  socialList: {
    type: Array as PropType<SocialInfo[]>,
  },
  socialBgColor: {
    type: String,
    default: '#68c8a2',
  },
  socialIconColor: {
    type: String,
    default: '#000',
  },
})
</script>

<template>
  <div class="a-card__box">
    <div
      class="a-card__content" :class="wrapClass" :style="{
        width,
      }"
    >
      <div class="a-card__head" :class="headerClass">
        <slot name="header">
          <img class="a-card__head--img" :src="avatar" alt="">
          <div class="a-card__head--name">
            {{ userName }}
          </div>
          <div class="a-card__head--role">
            {{ userRole }}
          </div>
        </slot>
      </div>
      <div class="a-card__info" :class="contactClass">
        <slot>
          <div v-for="item in contactList" :key="item.value" class="a-card__info--item">
            <i class="a-card__info--icon icon-btn" :class="item.icon" />
            <div class="a-card__info--right">
              <p class="a-card__info--main">
                {{ item.value }}
              </p>
              <p class="a-card__info--desc">
                {{ item.desc }}
              </p>
            </div>
          </div>
        </slot>
      </div>
      <div
        class="a-card__footer" :class="footerClass" :style="{
          background: socialBgColor,
          color: socialIconColor,
        }"
      >
        <slot name="footer">
          <a v-for="item in socialList" :key="item.link" :href="item.link" class="a-card__footer--item"> <i class="a-card__footer--icon icon-btn" :class="item.icon" /> </a>
        </slot>
      </div>
    </div>
  </div>
</template>

<style lang="less">
.a-card__box{
  padding-top: 2rem;
  padding-bottom: 2rem;
  .a-card__content {
    position: relative;
    width: 15rem;
    padding: 0 0.4rem;
    background-color: #fff;
    border-radius: 10px;
    box-shadow: 0 0 0.8rem 0.2rem rgba(0, 0, 0, 0.4);
    .a-card__head {
      position: relative;
      top: -1.6rem;
      padding: 1rem 2rem;
      text-align: center;
      color: #fff;
      background-color: #6bcca3;
      border-radius: 0.6rem;
      box-shadow: 0 0.2rem 0.4rem rgba(0, 0, 0, 0.4);
      &--img {
        display: block;
        margin: 0 auto;
        width: 4rem;
        height: 4rem;
        border-radius: 50%;
      }
      &--name {
        display: inline-block;
        font-size: 1.2rem;
        font-weight: bold;
      }
      &--role {
        font-size: 0.7rem;
      }
    }
    .a-card__info {
      margin-top: 1.2rem;
      &--item {
        display: flex;
        margin-top: 0.6rem;
        color: #61bc9f;
      }
      &--icon {
        display: inline-block;
        vertical-align: top;
        margin-right: 1rem;
        font-size: 1.2rem;
      }
      &--right {
        flex: 1;
        text-align: left;
      }
      &--main {
        font-size: 1rem;
        line-height: 1;
        color: #519f89;
        font-weight: bold;
      }
      &--desc {
        margin-top: 0.3rem;
        font-size: 0.7rem;
        line-height: 1;
      }
    }
    .a-card__footer {
      position: relative;
      top: 1rem;
      display: flex;
      margin-top: 1.8rem;
      padding: 0 0.8rem;
      color: #000;
      background-color: #68c8a2;
      border-radius: 0.4rem;
      box-shadow: 0 0.2rem 0.4rem rgba(0, 0, 0, 0.4);
      &--item {
        display: flex;
        align-items: center;
        justify-content: center;
        flex: 1;
        height: 3rem;
      }
      &--icon {
        display: inline-block;
        font-size: 1.4rem;
      }
    }
  }
}
</style>
