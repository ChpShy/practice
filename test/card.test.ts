import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import Card from '../src/components/Card.vue'

describe('Card.vue', () => {
  const contactList = ref([
    {
      icon: 'i-carbon-mobile',
      value: '5-555-666-777',
      desc: 'Mobile number',
    },
    {
      icon: 'i-carbon-email',
      value: 'example@gmail.com',
      desc: 'Email address',
    },
    {
      icon: 'i-carbon-flight-international',
      value: 'www.example.com',
      desc: 'Website address',
    },
  ])

  const socialList = ref([
    {
      icon: 'i-carbon-logo-facebook',
      link: 'https://www.facebook.com/',
    },
    {
      icon: 'i-carbon-logo-twitter',
      link: 'https://twitter.com/',
    },
    {
      icon: 'i-carbon-logo-instagram',
      link: 'https://www.instagram.com/',
    },
    {
      icon: 'i-carbon-logo-linkedin',
      link: 'https://www.linkedin.com/',
    },
  ])
  it('should render', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
      },
    })
    expect(wrapper.text()).toContain('Mobile App Developer')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('footer bg color should be changed', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        footerBgColor: '#ff0000',
      },
    })
    const element: any = wrapper.find('.a-card__footer').element
    expect(element.style.background).toBe('rgb(255, 0, 0)')
  })

  it('footer icon color should be changed', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        footerIconColor: '#ff0000',
      },
    })
    const element: any = wrapper.find('.a-card__footer').element
    expect(element.style.color).toBe('rgb(255, 0, 0)')
  })

  it('default slot should render', () => {
    const wrapper = mount(Card, {
      slots: {
        default: '<div>default content</div>',
      },
    })
    expect(wrapper.find('.a-card__info').element.innerHTML).toBe('<div>default content</div>')
  })

  it('header slot should render', () => {
    const wrapper = mount(Card, {
      slots: {
        header: '<div>header content</div>',
      },
    })
    expect(wrapper.find('.a-card__head').element.innerHTML).toBe('<div>header content</div>')
  })

  it('footer slot should render', () => {
    const wrapper = mount(Card, {
      slots: {
        footer: '<div>footer content</div>',
      },
    })
    expect(wrapper.find('.a-card__footer').element.innerHTML).toBe('<div>footer content</div>')
  })

  it('content class should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        wrapClass: 'content-class-test',
      },
    })
    expect(wrapper.find('.a-card__content').element.outerHTML).toContain('content-class-test')
  })

  it('content Object class  should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        wrapClass: {
          'content-class-test': true,
          'content-class-test-2': false,
        },
      },
    })
    expect(wrapper.find('.a-card__content').element.outerHTML).toContain('content-class-test')
    expect(wrapper.find('.a-card__content').element.outerHTML.includes('content-class-test-2')).toBeFalsy()
  })

  it('content Array class  should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        wrapClass: ['content-class-test', 'content-class-test-2'],
      },
    })
    expect(wrapper.find('.a-card__content').element.outerHTML).toContain('content-class-test')
    expect(wrapper.find('.a-card__content').element.outerHTML).toContain('content-class-test-2')
  })

  it('header class should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        headerClass: 'header-class-test',
      },
    })
    expect(wrapper.find('.a-card__head').element.outerHTML).toContain('header-class-test')
  })

  it('header object class should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        headerClass: 'header-class-test',
      },
    })
    expect(wrapper.find('.a-card__head').element.outerHTML).toContain('header-class-test')
    expect(wrapper.find('.a-card__head').element.outerHTML.includes('header-class-test-2')).toBeFalsy()
  })

  it('header Array class should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        headerClass: ['header-class-test', 'header-class-test-2'],
      },
    })
    expect(wrapper.find('.a-card__head').element.outerHTML).toContain('header-class-test')
    expect(wrapper.find('.a-card__head').element.outerHTML).toContain('header-class-test-2')
  })

  it('footer class should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        footerClass: 'footer-class-test',
      },
    })
    expect(wrapper.find('.a-card__footer').element.outerHTML).toContain('footer-class-test')
  })

  it('footer object class should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        footerClass: {
          'footer-class-test': true,
          'footer-class-test-2': false,
        },
      },
    })
    expect(wrapper.find('.a-card__footer').element.outerHTML).toContain('footer-class-test')
    expect(wrapper.find('.a-card__footer').element.outerHTML.includes('footer-class-test-2')).toBeFalsy()
  })

  it('footer Array class should add success', () => {
    const wrapper = mount(Card, {
      props: {
        avatar: 'https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image',
        userName: 'John Doe',
        userRole: 'Mobile App Developer',
        socialList: socialList.value,
        contactList: contactList.value,
        footerClass: ['footer-class-test', 'footer-class-test-2'],
      },
    })
    expect(wrapper.find('.a-card__footer').element.outerHTML).toContain('footer-class-test')
    expect(wrapper.find('.a-card__footer').element.outerHTML).toContain('footer-class-test-2')
  })
})
