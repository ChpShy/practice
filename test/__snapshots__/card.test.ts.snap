// Vitest Snapshot v1

exports[`Card.vue > footer slot should render 1`] = `
"<div class=\\"a-card__box\\">
  <div class=\\"a-card__content\\" style=\\"width: 15rem;\\">
    <div class=\\"a-card__head\\"><img class=\\"a-card__head--img\\" alt=\\"\\">
      <div class=\\"a-card__head--name\\"></div>
      <div class=\\"a-card__head--role\\"></div>
    </div>
    <div class=\\"a-card__info\\">
      <div>footer content</div>
    </div>
    <div class=\\"a-card__footer\\" style=\\"background: rgb(104, 200, 162); color: rgb(0, 0, 0);\\"></div>
  </div>
</div>"
`;

exports[`Card.vue > header slot should render 1`] = `
"<div class=\\"a-card__box\\">
  <div class=\\"a-card__content\\" style=\\"width: 15rem;\\">
    <div class=\\"a-card__head\\">
      <div>header content</div>
    </div>
    <div class=\\"a-card__info\\"></div>
    <div class=\\"a-card__footer\\" style=\\"background: rgb(104, 200, 162); color: rgb(0, 0, 0);\\"></div>
  </div>
</div>"
`;

exports[`Card.vue > should render 1`] = `
"<div class=\\"a-card__box\\">
  <div class=\\"a-card__content\\" style=\\"width: 15rem;\\">
    <div class=\\"a-card__head\\"><img class=\\"a-card__head--img\\" src=\\"https://p6-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/02ac8fd2cdb040f5bc7d73b562518d97~tplv-k3u1fbpfcp-watermark.image\\" alt=\\"\\">
      <div class=\\"a-card__head--name\\">John Doe</div>
      <div class=\\"a-card__head--role\\">Mobile App Developer</div>
    </div>
    <div class=\\"a-card__info\\">
      <div class=\\"a-card__info--item\\"><i class=\\"a-card__info--icon icon-btn i-carbon-mobile\\"></i>
        <div class=\\"a-card__info--right\\">
          <p class=\\"a-card__info--main\\">5-555-666-777</p>
          <p class=\\"a-card__info--desc\\">Mobile number</p>
        </div>
      </div>
      <div class=\\"a-card__info--item\\"><i class=\\"a-card__info--icon icon-btn i-carbon-email\\"></i>
        <div class=\\"a-card__info--right\\">
          <p class=\\"a-card__info--main\\">example@gmail.com</p>
          <p class=\\"a-card__info--desc\\">Email address</p>
        </div>
      </div>
      <div class=\\"a-card__info--item\\"><i class=\\"a-card__info--icon icon-btn i-carbon-flight-international\\"></i>
        <div class=\\"a-card__info--right\\">
          <p class=\\"a-card__info--main\\">www.example.com</p>
          <p class=\\"a-card__info--desc\\">Website address</p>
        </div>
      </div>
    </div>
    <div class=\\"a-card__footer\\" style=\\"background: rgb(104, 200, 162); color: rgb(0, 0, 0);\\"><a href=\\"https://www.facebook.com/\\" class=\\"a-card__footer--item\\"><i class=\\"a-card__footer--icon icon-btn i-carbon-logo-facebook\\"></i></a><a href=\\"https://twitter.com/\\" class=\\"a-card__footer--item\\"><i class=\\"a-card__footer--icon icon-btn i-carbon-logo-twitter\\"></i></a><a href=\\"https://www.instagram.com/\\" class=\\"a-card__footer--item\\"><i class=\\"a-card__footer--icon icon-btn i-carbon-logo-instagram\\"></i></a><a href=\\"https://www.linkedin.com/\\" class=\\"a-card__footer--item\\"><i class=\\"a-card__footer--icon icon-btn i-carbon-logo-linkedin\\"></i></a></div>
  </div>
</div>"
`;
